Makefile

```diff
@@ -431,7 +431,7 @@ push-operator-image: build-operator-image registry-login
 
 .PHONY: prepare-operator-manifests
 ## Prepare operator manifests
-prepare-operator-manifests: push-operator-image
+prepare-operator-manifests: build-operator-image
```

Comment:
It is not sufficient to just build the operator image here. You need to actually push the operator image to the repo in order to be able to get proper `OPERATOR_IMAGE_SHA` used later in couple of lines later.

That is why it needs to depend on `push-operator-image`

Reply:
Just to confirm: sha after building image locally and pushing to registry are not the same?

Reply:
# TL;DR

The digest by which you can pull image from repository is not image's own SHA, but the `Repo Digest` that is (apparently) created by the repository only when pushing the image. And so it is not sufficient to just build the image, you need to actually push the image to repo to get the digest.

# Showcase

## Start with no image

```sh
$ docker pull quay.io/pmacik/sbo-testing
Using default tag: latest
Error response from daemon: repository quay.io/pmacik/sbo-testing not found: name unknown: repository name not known to registry
```

## Build image

```sh
$ OPERATOR_REPO_REF=quay.io/pmacik/sbo-testing make build-operator-image

docker build -f Dockerfile.rhel -t quay.io/pmacik/sbo-testing:3c78276a .
Sending build context to Docker daemon  376.2MB
Step 1/29 : FROM openshift/origin-release:golang-1.14 AS builder
 ---> f7740938f0f7
...
Successfully built 9fb1d366f5bf
Successfully tagged quay.io/pmacik/sbo-testing:3c78276a
```

## Inspect image before push

```sh
$ docker inspect quay.io/pmacik/sbo-testing:3c78276a > docker-inspect-pre-push.json
```

## Show all SHAs before push

```sh
cat docker-inspect-pre-push.json| grep sha
        "Id": "sha256:9fb1d366f5bfb8a47263d44d413ba576581f34e8db163885879820a08b4f97b9",
        "Parent": "sha256:31379ed5dd5f732c0ca70b8687221d5aa7c700f56a5d4a0abb9489b020a7fd00",
            "Image": "sha256:31379ed5dd5f732c0ca70b8687221d5aa7c700f56a5d4a0abb9489b020a7fd00",
            "Image": "sha256:31379ed5dd5f732c0ca70b8687221d5aa7c700f56a5d4a0abb9489b020a7fd00",
                "sha256:d6ec160dc60fbfb962da0b216dfd27f00890bc8ace02994de6441a2517b04db5",
                "sha256:37330a2a1954a6b667ef15cab3299dc3cc82f5bc2bc2bbc46b1a6a938d3b325b",
                "sha256:7f8c34af1c42b8475892f3097a3d9f567adb6b497bb8474aae40b277b3ed9ba6"
```

## Actual images's SHA is the ID

```sh
cat docker-inspect-pre-push.json| jq -r '.[].Id'
```
or...

```sh
$ docker inspect --format='{{.Id}}' quay.io/pmacik/sbo-testing:3c78276a 
sha256:9fb1d366f5bfb8a47263d44d413ba576581f34e8db163885879820a08b4f97b9
```

## There is no repo digest yet

```sh
$ cat docker-inspect-pre-push.json| jq -r '.[].RepoDigests'
[]
```
or...

```sh
$ docker inspect --format='{{.RepoDigests}}' quay.io/pmacik/sbo-testing:3c78276a 
[]
```

## Now, lets push the image to a repo

```sh
$ docker push quay.io/pmacik/sbo-testing:3c78276a                               
The push refers to repository [quay.io/pmacik/sbo-testing]
7f8c34af1c42: Pushed 
37330a2a1954: Mounted from pmacik/servicebinding-operator 
d6ec160dc60f: Mounted from pmacik/servicebinding-operator 
3c78276a: digest: sha256:eeb71ea6f990b4ebad3391861e10f3380d190ce99405678382379a2d30cab9ed size: 949
```

The last line is the RepoDigest that we are looking for - let's check

## Inspect image after push

```sh
$ docker inspect quay.io/pmacik/sbo-testing:3c78276a > docker-inspect-after-push.json
```

## Check all SHAs after push

```sh
$ cat docker-inspect-after-push.json| grep sha                                             
        "Id": "sha256:9fb1d366f5bfb8a47263d44d413ba576581f34e8db163885879820a08b4f97b9",
            "quay.io/pmacik/sbo-testing@sha256:eeb71ea6f990b4ebad3391861e10f3380d190ce99405678382379a2d30cab9ed"
        "Parent": "sha256:31379ed5dd5f732c0ca70b8687221d5aa7c700f56a5d4a0abb9489b020a7fd00",
            "Image": "sha256:31379ed5dd5f732c0ca70b8687221d5aa7c700f56a5d4a0abb9489b020a7fd00",
            "Image": "sha256:31379ed5dd5f732c0ca70b8687221d5aa7c700f56a5d4a0abb9489b020a7fd00",
                "sha256:d6ec160dc60fbfb962da0b216dfd27f00890bc8ace02994de6441a2517b04db5",
                "sha256:37330a2a1954a6b667ef15cab3299dc3cc82f5bc2bc2bbc46b1a6a938d3b325b",
                "sha256:7f8c34af1c42b8475892f3097a3d9f567adb6b497bb8474aae40b277b3ed9ba6"
```

There is one additional - the RepoDigest

```sh
$ diff docker-inspect-pre-push.json docker-inspect-after-push.json 
8c8,10
<         "RepoDigests": [],
---
>         "RepoDigests": [
>             "quay.io/pmacik/sbo-testing@sha256:eeb71ea6f990b4ebad3391861e10f3380d190ce99405678382379a2d30cab9ed"
>         ],
```

## Now let's assume I'm deployment and want to pull the image from repo by image's `Id` - SHA: It does not work

```sh
$ docker pull quay.io/pmacik/sbo-testing@sha256:9fb1d366f5bfb8a47263d44d413ba576581f34e8db163885879820a08b4f97b9  
Error response from daemon: manifest for quay.io/pmacik/sbo-testing@sha256:9fb1d366f5bfb8a47263d44d413ba576581f34e8db163885879820a08b4f97b9 not found: manifest unknown: manifest unknown
```

## Let's try the correct SHA - RepoDigest: It works!

```sh
$ docker pull quay.io/pmacik/sbo-testing@sha256:eeb71ea6f990b4ebad3391861e10f3380d190ce99405678382379a2d30cab9ed
sha256:eeb71ea6f990b4ebad3391861e10f3380d190ce99405678382379a2d30cab9ed: Pulling from pmacik/sbo-testing
Digest: sha256:eeb71ea6f990b4ebad3391861e10f3380d190ce99405678382379a2d30cab9ed
Status: Image is up to date for quay.io/pmacik/sbo-testing@sha256:eeb71ea6f990b4ebad3391861e10f3380d190ce99405678382379a2d30cab9ed
quay.io/pmacik/sbo-testing@sha256:eeb71ea6f990b4ebad3391861e10f3380d190ce99405678382379a2d30cab9ed
```

## Summary

The digest by which you can pull image from repository is not image's own SHA, but the `Repo Digest` that is (apparently) created by the repository only when pushing the image. And so it is not sufficient to just build the image, you need to actually push the image to repo to get the digest.

Reply:
@pmacik thanks a lot for the excellent explanation! The change is reverted now.